File: 4 Algorithms on Strings/Week 2 Burrows-Wheeler Transform and Suffix Arrays/Programming-Assignment-2/bwt/bwt.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::vector;

string BWT(const string& text) {
  string result = "";
  vector<string> m(text.size());
  string newtext = text + text;
  
  for (int i = 0; i < m.size(); i++){
      m[i] = newtext.substr(i, i + text.size());
  }
  
  sort(m.begin(), m.end());
  
  for (int i = 0; i < m.size(); i++){
      result += m[i].substr(text.size() - 1 , 1);
  }

  return result;
}

int main() {
  string text;
  cin >> text;
  cout << BWT(text) << endl;
  return 0;
}

File: 4 Algorithms on Strings/Week 2 Burrows-Wheeler Transform and Suffix Arrays/Programming-Assignment-2/bwt/bwt.py
# python3
import sys

def BWT(text):
    result = ""
    n = len(text)
    m = [""] * n
    newtext = text + text
    
    for i in range(n):
        m[i] = newtext[i:i + n]
    m = sorted(m, key=str.lower)
    
    for i in range(n):
        result += m[i][-1]

    return result

if __name__ == '__main__':
    text = sys.stdin.readline().strip()
    print(BWT(text))